azkaban-exec-server/src/main/java/azkaban/dag/Flow.java

```diff
@@ -0,0 +1,141 @@
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package azkaban.dag;
+
+import static java.util.Objects.requireNonNull;
+
+import com.google.common.annotations.VisibleForTesting;
+import java.util.ArrayList;
+import java.util.List;
+
+class Flow {
+
+  private final String name;
+  private final FlowProcessor flowProcessor;
+  private final List<Node> nodes = new ArrayList<>();
+  private Status status = Status.READY;
+
+  Flow(final String name, final FlowProcessor flowProcessor) {
+    this.name = name;
+    requireNonNull(flowProcessor);
+    this.flowProcessor = flowProcessor;
+  }
+
+  Flow addNode(final Node node) {
+    node.setFlow(this);
+    this.nodes.add(node);
+    return this;
+  }
+
+  Flow addNodes(final Node... nodes) {
+    for (final Node node : nodes) {
+      addNode(node);
+    }
+    return this;
+  }
+
+  void start() {
+    changeStatus(Status.RUNNING);
+    final List<Node> readyNodes = getReadyNodes();
+    for (final Node node : readyNodes) {
+      node.run();
+    }
+    // It's possible that all nodes are disabled. In this rare case the flow should be
+    // marked success. Otherwise it will be stuck in the the running state.
+    checkFinished();
+  }
+
+  void kill() {
+    changeStatus(Status.KILLING);
+    for (final Node node : this.nodes) {
+      node.kill();
+    }
+    checkFinished();
+  }
+
+  /**
+   * Checks if the flow is done.
+   */
+  void checkFinished() {
+    // A flow may have nodes that are disabled. It's safer to scan all the nodes.
+    // The assumption is that the overhead is minimal. If it is not the case, we can optimize later.
+    boolean failed = false;
+    boolean killed = false;
+    for (final Node node : this.nodes) {
+      if (!node.isInTerminalState()) {
+        return;
+      }
+      switch (node.getStatus()) {
+        case FAILURE:
+          failed = true;
+          break;
+        case KILLED:
+          killed = true;
+          break;
+        default:
+          break;
+      }
+    }
+
+    if (failed) {
+      // If there are both failures and killed nodes, the flow status should be failed.
+      changeStatus(Status.FAILURE);
+    } else if (killed) {
+      changeStatus(Status.KILLED);
+    } else {
+      changeStatus(Status.SUCCESS);
+    }
+  }
+
+  /**
+   * Gets all the nodes that are ready to run.
+   *
+   * @return a list of nodes that are ready to run
+   */
+  private List<Node> getReadyNodes() {
+    final List<Node> readyNodes = new ArrayList<>();
+    for (final Node node : this.nodes) {
+      if (node.isReady()) {
```

Comment:
Just checking, I notice you have at least some unit test for sub-flows too, but I wonder if this works correctly? The thing is, you can't start a sub-flow "alone" even if it doesn't have any direct dependencies. It should be started once it's enclosing parent flow has been allowed to start. Maybe you are going to simplify this so that the parent flow node is added as a dependency to the starting nodes of a sub-flow?

Reply:
The sub-flows are handled at a higher level as demonstrated by the unit tests. 
At the DAG engine level, it doesn't have the sub-flow concept. This is to separate the concerns so that each layer's code can be simplified. 
The higher layer will make sure the sub-flow won't start until the parent flow tells it to start. This happens as the node representing the sub-flow is asked to run.